diff --git a/etc/vz.conf.in b/etc/vz.conf.in
index 5db7464..146ffaf 100644
--- a/etc/vz.conf.in
+++ b/etc/vz.conf.in
@@ -40,11 +40,9 @@ TEMPLATE=@VZDIR@/template
 ## Defaults for containers
 VE_ROOT=@VZDIR@/root/$VEID
 VE_PRIVATE=@VZDIR@/private/$VEID
-CONFIGFILE="vswap-256m"
-DEF_OSTEMPLATE="centos-6-x86"
 NAMESERVER=inherit # Copy from host system's /etc/resolv.conf
 ## Filesystem layout for new CTs: either simfs or ploop
-VE_LAYOUT=ploop
+#VE_LAYOUT=ploop
 
 # User namespace configuration
 LOCAL_UID=100000
